src/flow/src/expr/scalar.rs

```diff
@@ -59,3 +59,340 @@ pub enum ScalarExpr {
+
+    fn visit_mut_children<F>(&mut self, mut f: F)
+    where
+        F: FnMut(&mut Self),
+    {
+        match self {
+            ScalarExpr::Column(_)
+            | ScalarExpr::Literal(_, _)
+            | ScalarExpr::CallUnmaterializable(_) => (),
+            ScalarExpr::CallUnary { func, expr } => f(expr),
+            ScalarExpr::CallBinary { func, expr1, expr2 } => {
+                f(expr1);
+                f(expr2);
+            }
+            ScalarExpr::CallVariadic { func, exprs } => {
+                for expr in exprs {
+                    f(expr);
+                }
+            }
+            ScalarExpr::If { cond, then, els } => {
+                f(cond);
+                f(then);
+                f(els);
+            }
+        }
+    }
+}
+
+impl ScalarExpr {
+    /// if expr contains function `Now`
+    pub fn contains_temporal(&self) -> bool {
+        let mut contains = false;
+        self.visit_post_nolimit(&mut |e| {
+            if let ScalarExpr::CallUnmaterializable(UnmaterializableFunc::Now) = e {
+                contains = true;
+            }
+        });
+        contains
+    }
+
+    /// extract lower or upper bound of `Now` for expr, where `lower bound <= expr < upper bound`
+    ///
+    /// returned bool indicates whether the bound is upper bound:
+    ///
+    /// false for lower bound, true for upper bound
+    /// TODO(discord9): allow simple transform like `now() + a < b` to `now() < b - a`
+    pub fn extract_bound(&self) -> Result<(Option<Self>, Option<Self>), String> {
+        let unsupported_err = |msg: &str| {
+            Err(format!(
+                    "Unsupported temporal predicate: {msg}. NOTE: Use `now()` in direct comparison: {:?}",
+                    self
+                ))
+        };
+        if let Self::CallBinary {
+            mut func,
+            mut expr1,
+            mut expr2,
+        } = self.clone()
+        {
+            if !(expr1.contains_temporal() ^ expr2.contains_temporal()) {
+                return unsupported_err("one side of the comparison must be `now()`");
+            }
+            if !expr1.contains_temporal()
+                && *expr2 == ScalarExpr::CallUnmaterializable(UnmaterializableFunc::Now)
+            {
+                std::mem::swap(&mut expr1, &mut expr2);
+                func = match func {
+                    BinaryFunc::Eq => BinaryFunc::Eq,
+                    BinaryFunc::NotEq => BinaryFunc::NotEq,
+                    BinaryFunc::Lt => BinaryFunc::Gt,
+                    BinaryFunc::Lte => BinaryFunc::Gte,
+                    BinaryFunc::Gt => BinaryFunc::Lt,
+                    BinaryFunc::Gte => BinaryFunc::Lte,
+                    _ => {
+                        return unsupported_err("The top level operator must be comparison");
+                    }
+                };
+            }
+            // TODO: support simple transform like `now() + a < b` to `now() < b - a`
+            if expr2.contains_temporal()
```

Comment:
1. We can refactor this code to make it more readable
2. `contains_temporal` is a potentially heavy calculation, it is recommended to assign it to a variable